pkg/k8sops/generators.go

```diff
@@ -0,0 +1,307 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package k8sops
+
+import (
+	"fmt"
+
+	m3dboperator "github.com/m3db/m3db-operator/pkg/apis/m3dboperator"
+	myspec "github.com/m3db/m3db-operator/pkg/apis/m3dboperator/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	_probeTimeoutSeconds      = 30
+	_probeInitialDelaySeconds = 10
+	_probeFailureThreshold    = 15
+	_probePort                = 7201
+	_probePath                = "/health"
+
+	_baseImage              = "quay.io/m3/m3dbnode:latest"
+	_dataDirectory          = "/var/lib/m3db/"
+	_configurationDirectory = "/etc/m3db/"
+	_configurationName      = "m3-configuration"
+	_configurationFileName  = "m3.yml"
+)
+
+var (
+	_configurationFileLocation = fmt.Sprintf("%s%s", _configurationDirectory, _configurationFileName)
+)
+
+// GenerateCRD generates the crd object needed for the M3DBCluster
+func (k *k8sops) GenerateCRD() *apiextensionsv1beta1.CustomResourceDefinition {
```

Comment:
I think we might want to add a Status subresource here. There's a helpful blog post about it: https://blog.openshift.com/kubernetes-custom-resources-grow-up-in-v1-10/

It only targets 1.10 so we'd have to check the cluster version, but it seems like the idiomatic way to do CRD statuses in 1.10+. Maybe we only want the operator to even target those versions?